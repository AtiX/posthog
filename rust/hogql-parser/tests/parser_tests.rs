// tests/parser_tests.rs
use hogql_parser::parse_query;
use insta::assert_json_snapshot;
use std::fs;
use std::path::Path;

#[test]
fn test_queries() {
    let queries_dir = Path::new("tests/queries");
    for entry in fs::read_dir(queries_dir).expect("Failed to read queries directory") {
        let entry = entry.expect("Failed to read directory entry");
        let path = entry.path();
        if path.extension().and_then(|s| s.to_str()) == Some("sql") {
            let query_name = path.file_stem().and_then(|s| s.to_str()).unwrap();
            let query = fs::read_to_string(&path).expect("Failed to read query file");
            let ast = parse_query(&query);
            assert_json_snapshot!(query_name, ast);
        }
    }
}

#[test]
fn test_select_with_and() {
    let query = "SELECT * FROM users WHERE active = true AND role = 'admin';";
    let ast = parse_query(query);
    assert_json_snapshot!(ast);
}

#[test]
fn test_select_with_or() {
    let query = "SELECT id FROM orders WHERE status = 'pending' OR status = 'processing';";
    let ast = parse_query(query);
    assert_json_snapshot!(ast);
}

#[test]
fn test_invalid_query() {
    let query = "SELECT FROM WHERE;";
    let result = std::panic::catch_unwind(|| parse_query(query));
    assert!(result.is_err(), "Parser should panic on invalid query");
}
