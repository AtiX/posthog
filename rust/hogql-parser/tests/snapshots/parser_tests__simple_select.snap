---
source: hogql-parser/tests/parser_tests.rs
expression: ast
---
{
  "Select": {
    "columns": [
      {
        "Identifier": "*"
      }
    ],
    "from": {
      "Identifier": "users"
    },
    "where_clause": {
      "BinaryOp": {
        "left": {
          "BinaryOp": {
            "left": {
              "Identifier": "active"
            },
            "op": "=",
            "right": {
              "Identifier": "true"
            }
          }
        },
        "op": "AND",
        "right": {
          "BinaryOp": {
            "left": {
              "Identifier": "role"
            },
            "op": "=",
            "right": {
              "Literal": "admin"
            }
          }
        }
      }
    }
  }
}
